docs/codeql/codeql-language-guides/analyzing-data-flow-in-go.rst

```diff
@@ -0,0 +1,369 @@
+
+.. code-block:: ql
+
+   import go
+
+   module MyFlowConfiguration implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       source instanceof RemoteFlowSource
+     }
+
+     ...
+   }
+
+   module MyTaintFlow = TaintTracking::Global<MyFlowConfiguration>;
+
+Exercises
+~~~~~~~~~
+
+Exercise 2: Write a query that finds all hard-coded strings used to create a ``url.URL``, using global data flow. (`Answer <#exercise-2>`__)
+
+Exercise 3: Write a class that represents flow sources from ``os.Getenv(..)``. (`Answer <#exercise-3>`__)
+
+Exercise 4: Using the answers from 2 and 3, write a query which finds all global data flows from ``os.Getenv`` to ``url.URL``. (`Answer <#exercise-4>`__)
+
+Answers
+-------
+
+Exercise 1
+~~~~~~~~~~
+
+.. code-block:: ql
+
+   import go
+
+   from Function urlParse, Expr arg, StringLit rawURL, CallExpr call
+   where
+     (
+       urlParse.hasQualifiedName("url", "Parse") or
+       urlParse.hasQualifiedName("url", "ParseRequestURI")
+     ) and
+     call.getTarget() = urlParse and
+     arg = call.getArgument(0) and
+     DataFlow::localFlow(DataFlow::exprNode(rawURL), DataFlow::exprNode(arg))
+   select call.getArgument(0)
+
+Exercise 2
+~~~~~~~~~~
+
+.. code-block:: ql
+
+   import go
+
+   module LiteralToURLConfig implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       source.asExpr() instanceof StringLit
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       exists(Function urlParse, CallExpr call |
+         (
+           urlParse.hasQualifiedName("url", "Parse") or
+           urlParse.hasQualifiedName("url", "ParseRequestURI")
+         ) and
+         call.getTarget() = urlParse and
+         sink.asExpr() = call.getArgument(0)
+       )
+     }
+   }
+
+   module LiteralToURLFlow = DataFlow::Global<LiteralToURLConfig>;
+
+   from DataFlow::Node src, DataFlow::Node sink
+   where LiteralToURLFlow::flow(src, sink)
+   select src, "This string constructs a URL $@.", sink, "here"
+
+Exercise 3
+~~~~~~~~~~
+
+.. code-block:: ql
+
+   import go
+
+   class GetenvSource extends CallExpr {
+     GetenvSource() {
+       exists(Function m | m = this.getTarget() |
+         m.hasQualifiedName("os", "Getenv")
+       )
+     }
+   }
+
+Exercise 4
+~~~~~~~~~~
+
+.. code-block:: ql
+
+   import go
+
+   class GetenvSource extends CallExpr {
+     GetenvSource() {
+       exists(Function m | m = this.getTarget() |
+         m.hasQualifiedName("os", "Getenv")
+       )
+     }
+   }
+
+   module GetenvToURLConfig implements DataFlow::ConfigSig {
+     predicate isSource(DataFlow::Node source) {
+       source instanceof GetenvSource
+     }
+
+     predicate isSink(DataFlow::Node sink) {
+       exists(Function urlParse, CallExpr call |
+         (
+           urlParse.hasQualifiedName("url", "Parse") or
+           urlParse.hasQualifiedName("url", "ParseRequestURI")
+         ) and
```

Comment:
```suggestion
         urlParse.hasQualifiedName("url", ["Parse", "ParseRequestURI"]) and
```

Reply:
I deliberately didn't do this because I think it's a bit harder to understand. I think this guide should just be a simple approach that is easy to understand and which works.

Reply:
You could mention it as a note underneath the example. "For brevity, we could also shorten `...` to `...`".

Reply:
To my mind, the aim of this guide is not to give the best way to write things, but to help the reader use data flow and to be as clear as possible. I think that introducing a new notation to do the same thing  does not help with that, especially when it isn't very easy to see at a glance what it is doing.